refactor to new architecture

pleimer opened the pull request:
This PR is for tracking the final re-factoring work into the new architecture. At the time of posting, the re-factored project satisfies the original functionality of sg-core, albeit with different build, configure and execute processes. The general overview can be found in the main [README](https://github.com/pleimer/sg-core/blob/refactor/README.md). Documentation for plugin development can be found in the [docs folder](https://github.com/pleimer/sg-core/blob/refactor/docs/developer/developing-plugins.md). For review, it is better just to look at the processes outlined in the markdown docs and view each [plugin's code](https://github.com/pleimer/sg-core/tree/refactor/plugins) and comment/make suggestions on those.

I will be continuing to make updates to this PR, thus certain interfaces will continue to change. However the metrics path is 99% finished and may be only tweaked before I try to merge this. 

Additionally, end to end performance testing still needs to be done on the fastpath.

leifmadsen commented on README.md:
```diff
@@ -1,17 +1,102 @@
+config block specific to that plugin.
+
+## Example Configuration
+This configuration assumes both a Quipid Dispatch Router and Prometheus instance
```
```suggestion
This configuration assumes both a QPID Dispatch Router and Prometheus instance
```

pleimer replied:
I remember thinking that looked weird when writing that sentence

pleimer commented:
My initial architectural brainstorming can be found here: https://docs.google.com/document/d/10HupzNlT9J4AeIyy4qKmvsMwU0RLsLRczVmmQpm8awI/edit

This includes diagrams and example configurations. For convenience, here is the general architecture diagram for how Transport, Handler, and Application plugin are managed by the sg-core: 


![arch1(4)](https://user-images.githubusercontent.com/41337120/104632195-5acff680-566b-11eb-836e-3ce044020038.png)

atyronesmith commented:
Should the Application be inside the sg-core dotted box?

pleimer commented:
@atyronesmith application plugins interact with programs like Prometheus. For example the [Prometheus Plugin](https://github.com/pleimer/sg-core/tree/refactor/plugins/application/prometheus) application plugin prepares metrics for storage in Prometheus.

atyronesmith commented on plugins/application/prometheus/expiry.go:
```diff
@@ -0,0 +1,60 @@
+func newExpiryProc() *expiryProc {
+	return &expiryProc{
+		entries:  list.New(),
+		interval: 10000,
```
Expire time might need to be based on the individual metric's sample interval.  I believe if the expire interval is set to 10 seconds and a variable's collectd interval is 60 seconds, there code be issues with variables expiring before they are sampled by prometheus.  The prometheus sample interval should be set to 60 seconds.  It is possible that a variable will get set and expire before it is sampled and then the timeseries will go invalid inside Prometheus.  

This brings up the issue of putting all metrics of the same interval into their own scrape target so that Prometheus can poll at different rates.  For example, disk-related metrics only need to be sample every 10 minutes (as an example), but packet counts every 30 seconds.  With a single scrape target, Prometheus would
need to scrape every 30 seconds thereby oversampling the disk metrics.  The solution is to have bins of scrape targets grouped by collectd interval.  The collectd interval is contained in each metric that is sent.

:)

pleimer replied:
This configuration refers to how often the expiry process checks registered metrics to see if they are expired, not the actual interval for metrics: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/expiry.go#L55

Currently a metric is marked as stale if one does not arrive in the timeout interval specified in the Prometheus Plugin config: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/main.go#L239. By default this is 20s: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/main.go#L168. Currently that configured timeout applies to all incoming metrics, but it would be simple to use an interval specified by the metric itself. I definitely see the utility in batching metrics according to timeout and seems like functionality we should have, I will add that to my backlog. That is not what happens in the current sg-core, however. Thanks for the feedback and ideas!

atyronesmith replied:
Correct.  Adding time-based bins would be an enhancement.   However, the stale interval should really be based on individual metrics.  This was a glaring problem with the original implementation.

atyronesmith commented:
> @atyronesmith application plugins interact with programs like Prometheus. For example the [Prometheus Plugin](https://github.com/pleimer/sg-core/tree/refactor/plugins/application/prometheus) prepares metrics for storage in Prometheus.

Ok.  I understand what is meant by Application now :).

pleimer commented on docs/developer/developing-plugins.md:
````diff
@@ -0,0 +1,148 @@
+
+```go
+func (t *TCP) Run(ctx context.Context, wg *sync.WaitGroup, w transport.WriteFn, done chan bool) transport.Transport {
+    defer wg.Done()
````
I will probably add this functionality to a decorator function so that plugin developers do not need to worry about it.

csibbitt commented on plugins/transport/socket/main.go:
```diff
@@ -0,0 +1,90 @@
+			//fmt.Printf("received message: %s\n", string(msgBuffer))
+
+			if err != nil || n < 1 {
+				done <- true
```
Looking to see if anything generates an error in this case. Does this survive multiple open/closes from the bridge?

pleimer replied:
I'm haven't tested it. This is the condition used by the original sg-core

atyronesmith replied:
This needs to be wrapped in error handling code for socket closures as Chris mentioned...

csibbitt reviewed (commented):
I wouldn't mind another pass over this and the chance for others to comment, but it looks really good, thanks Paul!

Before merging this to master, I think it's worth considering timing vs changes required in the operator to support the new config interface, but otherwise I don't have any objections.

paramite commented on docs/developer/developing-plugins.md:
```diff
@@ -0,0 +1,148 @@
+
+Handlers parse incoming blobs from the transport into objects and delivers those objects to the internal buses. There are two types of handlers: metric handlers and event handlers. Metric handlers deliver metric objects to the internal metrics bus while event handlers deliver event objects to the internal events bus. These metrics and events are then consumed by the application plugins.
+
+Handlers must remain simple: they take in no configuration and should only handle one message type. See the [collectd-metrics](https://github.com/pleimer/sg-core-refactor/tree/master/plugins/handler/collectd-metrics) plugin for an example. Additionally, handlers should not print logs (this is why no logger is passed into the New() function). If errors occur while parsing messages, handlers should ...
```
Kind of agree with Chris here. I don't see advantage of generating logs in handler itself too. There will have to be logic in each application processing the specific handler messages. If author of any application will forget to handle errors (or simply will not care), the operators might loose important information about failures. I'm not saying that handlers should not send error messages, just that failures should IMO be logged also on handler layer.

pleimer replied:
I see the concern, it definitely leaves information out. The reason I chose to suggest against it is that because handlers are dealing with a large quantity of messages, the logs could become extremely unwieldy very quickly. If the developer could be responsible enough to log the first error, and all proceeding duplicates of the error ignore, maybe it is worth it. What do you guys think?

@paramite In most cases I don't think the Application should contain logic to deal with specific metrics but rather should only contain the logic to deal with the [metric type](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/pkg/data/data.go#L28). For events, maybe so if the plugin needs to react to them. In the case of parser errors in the collectd-metrics handler, the handler creates a new metric with a counter specifying the number of parsing errors that have occurred [here](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/plugins/handler/collectd-metrics/main.go#L53). The Prometheus plugin doesn't do anything special with this metric and instead just stores it like all the rest. But, we can plot this metric in Grafana and see if any parsing errors are occurring without filling up the logs.

pleimer replied:
Maybe the handler could return an error in this case and the manager could then unbind it from the transport. This may be the better option and wouldn't leave so much up to the plugin developer

paramite commented on plugins/handler/collectd-metrics/main.go:
```diff
@@ -0,0 +1,153 @@
+				Name:  genMetricName(cdmetric, index),
+				Type:  strToMetricType(cdmetric.Dstypes[index]),
+				Value: cdmetric.Values[index],
+				Time:  cdmetric.Time.Time(),
```
We have IIRC.

pleimer replied:
I think you are right

paramite commented on cmd/manager/manager.go:
```diff
@@ -0,0 +1,196 @@
+		mChan := make(chan []data.Metric)
+
+		eventBus.Subscribe(eChan)
+		metricBus.Subscribe(mChan)
```
From my reading the cost is speed. AFAIR the more readers on a channel the slower it gets. I didn't do any benchmarking though, can find the article about it I guess. We might not see any performance issues currently, but if will will, subscribing only where is necessary is first thing which might help with that.

atyronesmith replied:
I would suggested only subscribing to the bus(es) you need to use.  The other consideration would be to use buffered channels and send in bulk (i.e. group multiple metrics into a single chunk).

pleimer replied:
Having multiple processes lread on one channel isn't an issue here. There is only ever one application process listening on any one channel at a time, since the [bus object](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/pkg/bus/bus.go#L39) accepts a new channel object every time Subscribe() is called. The bus object does a multicast on all of the subscribed channels whenever Publish() is called. Run() is still implemented poorly. Switching between channels can cause slowdowns if a lot of messages are coming from both and it still enables the problem of accidentally locking up one of the buses if the plugin forgets to listen on both. A buffered channel would eventually fill up and block in this circumstance as well. This was one of those things that I said "lll deal with later" and forgot about :). 

What could be done is give the Application plugin both the metric and event bus and let it decide which to subscribe to. The only reason I did not do this earlier was because I was not sure that the Applications should have to understand the internal bus structure, but it really doesn't matter. Or, instead of giving the plugins channels to listen on, the manager could expect the implementation of a separate callback for metrics and events. If I did this, it may be worth switching the bus structure over to callbacks as well.

atyronesmith replied:
Please see my comment in bus.go about non-blocking sends... @pleimer

paramite commented on cmd/manager/manager.go:
```diff
@@ -0,0 +1,196 @@
+		}
+
+		var hType string
+		switch new := n.(type) {
```
Overriding builtin function new is never a good idea IMO :).

pleimer replied:
I was thinking back to my c++ days when overloading functions was acceptable :). The reason I did that here was just so that the configuration file would not need to separate handlers into separate sections for metrics and events

paramite commented on cmd/manager/manager.go:
```diff
@@ -0,0 +1,196 @@
+	return nil
+}
+
+//RunTransports spins off tranpsort + handler processes
```
s/tranpsort/transport/

paramite commented on plugins/application/prometheus/main.go:
```diff
@@ -0,0 +1,273 @@
+		case <-ctx.Done():
+			goto done
+		case <-eChan:
+			p.logger.Warn("Prometheus plugin received an event - disregarding")
```
This has to be run. Otherwise the channel gets blocking and whole loop on the "write side" gets stuck.

paramite reviewed (commented):
Few comments/nits added. Will have to give it one more round and actually run the code. I really love the architecture btw.

pleimer commented on pkg/concurrent/concurrent.go:
```diff
@@ -0,0 +1,77 @@
+package concurrent
```
You know I used that library before, but for some reason I didn't remember that it existed and I didn't see it in my brief search. Thanks for the reminder!

pleimer commented on plugins/handler/collectd-metrics/main_test.go:
```diff
@@ -0,0 +1,110 @@
+
+func BenchmarkParsing(b *testing.B) {
+	// GOMAXPROCS = 8
+	// On thinkpad T480s, performs at ~ 195k m/s
```
End to end is the next step!

pleimer commented on cmd/main.go:
```diff
@@ -0,0 +1,114 @@
+	"syscall"
+
+	"github.com/infrawatch/apputils/logging"
+	log "github.com/infrawatch/apputils/logging"
```
definitely not. I must have forgotten that I made the alias

pleimer commented on cmd/main.go:
```diff
@@ -0,0 +1,114 @@
+	"github.com/infrawatch/apputils/logging"
+	log "github.com/infrawatch/apputils/logging"
+	"github.com/infrawatch/apputils/system"
+	"github.com/infrawatch/sg-core-refactor/cmd/manager"
```
oof that's the trouble with starting in a different repo. I am sure there are other instances

pleimer commented on cmd/manager/manager.go:
```diff
@@ -0,0 +1,196 @@
+	metricHandlers = map[string][]handler.MetricHandler{}
+	eventHandlers = map[string][]handler.EventHandler{}
+	applications = map[string]application.Application{}
+	pluginPath = "/usr/lib64/sg-core"
```
It is! https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/cmd/config.go#L8

pleimer commented on docs/developer/developing-plugins.md:
```diff
@@ -0,0 +1,148 @@
+
+## Transports
+
+Transport plugins listen on an external messaging protocal and deliver received messages to handlers that have been bound to it per the administrator's configuration. They receive a configuration block from the main configuration file and write to handlers by involing transport.WriteFn in Run().
```
also s/involing/invoking

pleimer commented on docs/developer/developing-plugins.md:
````diff
@@ -0,0 +1,148 @@
+
+Handler plugin objects must imeplement either the MetricHandler interface or the EventHandler interface:
+```go
+type MetricHandler interface {
````
EventHandler has not been fleshed out yet since the event path has not fully been implemented. This could change, especially based on our conversation above

pleimer commented on pkg/bus/bus.go:
```diff
@@ -0,0 +1,60 @@
+
+/* TODO: optimize this
+channels are not particularily fast, best to avoid them in the fastpath and use Mutexes instead
+benchmark implementation with callback vs channels here
```
I ran the benchmark for the 2 out of 3 "stages" that happen in the sg-core: parsing the message and sending the message to the application plugins. Based on those, it seems we have some headroom for the 100km/s goal. End to end testing is still absolutely necessary and I plan to do that as soon as I get the SGO and STO updates far enough

pleimer commented on plugins/application/prometheus/main.go:
```diff
@@ -0,0 +1,273 @@
+}
+
+//PromCollector implements prometheus.Collector for incoming metrics. Metrics
+// with differing label dimensions must create separate PromCollectors.
```
Here are the points where the Prometheus golang library runs the checks:
[1](https://github.com/dnesting/client_golang/blob/449ccefff16c8e2b7229f6be1921ba22f62461fe/prometheus/registry.go#L394)
[2](https://github.com/dnesting/client_golang/blob/449ccefff16c8e2b7229f6be1921ba22f62461fe/prometheus/registry.go#L462)
[3](https://github.com/dnesting/client_golang/blob/449ccefff16c8e2b7229f6be1921ba22f62461fe/prometheus/registry.go#L475)
[4](https://github.com/dnesting/client_golang/blob/449ccefff16c8e2b7229f6be1921ba22f62461fe/prometheus/registry.go#L485)

Each calls this [checkConsistency()](https://github.com/dnesting/client_golang/blob/449ccefff16c8e2b7229f6be1921ba22f62461fe/prometheus/registry.go#L516) method. AFAICT, it only makes sure that descriptors and metric labels align, not that a collector contains metrics with all the same labels. Might still be worth running it to find out

pleimer commented on plugins/handler/collectd-metrics/main.go:
```diff
@@ -0,0 +1,153 @@
+		metrics = append(metrics, ms...)
+	}
+
+	metrics = append(metrics, []data.Metric{{
```
Yeah that is what I was referring to.

atyronesmith replied:
Hmmm...  Are inline metrics generated here because the handlers are now separate executables from the Applications?  While this traffic is internal, it could still greatly multiple the number of messages, encoding/decoding.  Depending on the bundling ratio from collectd, the number of sg_total_metric_rcv_count messages could equal the number of metrics sent per period.  It would be acceptable to only include these counters once a second in the stream. (i.e. a second optional append)

pleimer replied:
As we discuss this more, I am beginning to think the internal bus should consist of a large cache of metric objects only created once and then updated by the handlers. It would save the overhead of object allocation and provide effectively the same functionality except that creating the metric would be a function call rather than a object creation. Expiration of metrics could be handled by the bus then. The events may need to stay the same since they are more variable

atyronesmith commented on pkg/bus/bus.go:
```diff
@@ -0,0 +1,60 @@
+			c <- e
+		}(c, e)
+	}
+	eb.rw.RUnlock() //defer is actually very slow
```
I believe the go func at line 32 means that you won't block, but you could possibly see a crash if you run out of go routine space.  With the sg-bridge, overflows are treated as part of the communication channel (i.e. lost udp packet). For metrics, discarding is fine as there should be another metric coming along soon enough.  The sg-bridge uses a ring buffer so that the oldest metric is thrown away.  Can't really do that here.  What about using a non-blocking send?  i.e.

```
for _, c := range eb.subscribers {
  select {
      case c <- e:
        sent++
      default:
         dropped++
      }
}

The channels should be buffered to allow for goroutine scheduling.
```

atyronesmith replied:
Same for metrics below...

atyronesmith commented on docs/developer/developing-plugins.md:
````diff
@@ -0,0 +1,148 @@
+Application plugins must implement the Application interface:
+```go
+type Application interface {
+	Config([]byte) error
````
Can the Config contain whether the apps wants a metric and/or an event bus?  The manager would then be able to supply the appropriate channel(s).

pleimer replied:
Yeah that's a good idea. That way, the admin can choose which messages should be received by the application.

atyronesmith commented on cmd/config.go:
```diff
@@ -0,0 +1,29 @@
+	} `validate:"dive"`
+	Applications []struct {
+		Name   string `validate:"required"`
+		Config interface{}
```
Metrics bool
Events bool

pleimer commented:
@paramite the last PR fixes the issue you had with WaitGroups

pleimer commented:
@paramite this last PR fixes the issues you had with WaitGroups and should prevent anyone else from making the same mistake

pleimer commented on plugins/application/print/main.go:
```diff
@@ -0,0 +1,101 @@
+
+//Config implements application.Application
+func (print *Print) Config(c []byte) error {
+	print.configuration = configT{
```
@paramite just a heads up, you can set defaults by initializing the structure with values. They are overwritten if a value is provided by the config file

paramite replied:
Theoretically this can work, but practically it does not:

2021-01-25 11:38:02 [ERROR] failed to open metrics data output file [error: open : no such file or directory, plugin: print]
2021-01-25 11:38:02 [ERROR] failed to open events data output file [plugin: print, error: open : no such file or directory]

paramite replied:
After revert of this commit:
2021-01-25 12:11:56 [INFO] writing processed data to files. [plugin: print, events: /dev/stdout, metrics: /dev/stdout]

pleimer replied:
huh I'm suprised, I made that work in other instances of using this method for config files.

paramite reviewed (approved):
I've been able to run this with all stack pieces (except Prometheus since I'm focusing on events instead). So this is good enough for me currently. IMO we can fix all the issues when they arise instead of adding to this already huge PR. Approving for merge, but we should create stable branch for old version before.

pleimer commented:
Let me complete changes in SGO and STO for this first and run at least one performance test

pleimer commented:
The patch I just merged has the following description:


Here is the results of extensive performance testing and optimization of the metric path for the new architecture. For the performance testing, I used the following CRC environment with STF and monitoring enabled:

CodeReady Containers version: 1.17.0+99f5c87
OpenShift version: 4.5.14 (embedded in binary)
CPUs: 24
Memory: 65536MiB
Disk: 130GiB

Each test generated 100 million metrics in total.

Results:
**Build** | **Process Rate (msg / s)** | Accuracy
-|-|-
Original | 94.8k | 99.8%
Refactored | 71.5k | 99.8%
Refactored Optimized (this PR) | 91.6k | 99.8%

The build from this PR runs -3.5% slower than the original sg-core. A small slowdown was expected due to the increased complexity of the new architecture. Achieving within 5% of the original performance was the goal and that has been achieved.

The build from this PR runs 22% faster than the originally proposed refactor.  Most optimizations entailed minimizing the amount of allocations being made in the fast path as well as utilizing the ideal case of the sync.Map object for caching.

There are a number of changes to the framework that had to be made in order to achieve this improvement:

### 1. Metric bus changes
The metric bus no longer utilizes channels to broadcast from the metric handler to the applications. Rather, a callback function is used. There are a number of advantages to this approach that I believe outweigh the disadvantages.
Advantages:
- minimal object allocation
After an in depth CPU profiling of the system, it came to my attention that object allocation was taking a significant amount of time. Part of that reason was that for each incoming metric, a data.Metric object was created and sent to the application over this bus. In this new approach, the metric interface is enforced by a function definition with the arguments building up the body of the metric. In this way, the runtime does not have to allocate as much memory every time a metric is made.
- less chance for data corruption in the internal bus. 
If the metric sends a slice of metric objects to the application plugins, the data in the slice can be overwritten before the receiving side gets a chance the process the object and data corruption occurs (slices and channels can be dangerous in general). To mitigate this in the original refactor, the code allocated a new slice every time a message needed to be sent. As you can imagine, this was expensive.

With this new approach, fewer objects are allocated by the bus and since functions rather than data objects are used to to send information (and that information is mostly in primitive form), there is less chance for data corruption.

Disadvantages:
- long write and receive function signatures
- Application plugin writers must keep the processing time of receive functions short else it will slow down the whole sg-core rather than just the plugin process.

### 2. Application interface changes

I did not flesh out the event interface for applications yet because I figured it would be best to do that when we merge the events code into this. I created two new types for application plugins: the MetricReceiver and the EventReceiver. Applications can implement one, both, or none of the receiver interfaces if they so chose (in the latter case, it won't be listening on the bus). Warnings are thrown by the manager if a plugin developer forgets to implement one of those in their application plugin.

These subtypes define the function signature for the receive function that is called by the bus when new data arrives. Clearly, the processing must be kept short to keep speed. In the Prometheus plugin, incoming metrics are updated in a cache that is later scrapped. Since our use case is basically the ideal use case for sync.Map, I used that for caching and it performs extremely well. The only caveat is that the advantages of using that object are not seen at lower core counts (< 5), so we should keep the recommendation to at least 6 cores. [Here's a good write-up about the topic](https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c) 

### 3. Metric handler interface changes

I additionally added a new Run() function to the MetricHandler interface. This is not for running the main metric parsing and write-to-bus functionality that the Handle() function satisfies, but rather to provide an ongoing process for the handler to write its own metrics to the bus without taking up time on the fast path. Now, the collectd-metrics handler utilizes the Run() function to write internal metrics like `sg_total_msg_received_count` and `sg_total_metric_decode_count` to the bus at one second intervals. To have these written to the bus every time Handle() was called caused a fairly drastic reduction in speed. 

## Other changes
I also went ahead and added metric expiration based on the interval included in the message and a timestamp enable/disable configuration to the Prometheus plugin.

## Conclusion
This is a fairly large change from what I originally proposed so we will have to figure out how to merge events changes with this. I don't think it should be too difficult. I suggest we move the event bus (and logging bus) to a similar architecture if it isn't too inconvenient. Though the speed of the event bus is not as important, it would be good for consistency.

paramite commented on pkg/application/application.go:
```diff
@@ -11,5 +12,6 @@ import (
 type Application interface {
 	Config([]byte) error
-	Run(context.Context, chan data.Event, chan []data.Metric, chan bool)
+	RecieveMetric(string, time.Time, data.MetricType, time.Duration, float64, []string, []string)
```
nit: Receive

pleimer replied:
thanks!

pleimer commented on pkg/bus/bus.go:
```diff
@@ -35,26 +36,33 @@ func (eb *EventBus) Publish(e data.Event) {
-		go func(c chan []data.Metric, m []data.Metric) {
-			c <- m
-		}(c, m)
+func (mb *MetricBus) Publish(name string, time time.Time, typ data.MetricType, interval time.Duration, value float64, labelKeys []string, labelVals []string) {
```
Agreed, it is a lot uglier than the original which is unfortunate. The original design used the data.Metric type, which had all of the same fields within it as the function signature for the callback. I tried to standardize it so that metrics from pretty much anywhere would fit in. So: name field, time stamp, interval, value, and labels. Metadata fields may be worth adding. Ceilometer metrics is next so if I find that necessary I will add it.

I didn't try any third party queues. Anything that involves using the network stack of the OS I didn't even consider testing because I don't think there is any way that would be faster than keeping everything at a lower level within the golang runtime and thus did not take the time to implement and test. There would have to be additional processing done to re-marshal the incoming message into a byte string representation of the data just to send it to another sg-core component that would then have to unmarshal it again. That would be significantly more overhead, and based on my cpu profiling, the unmarshalling already takes up a fair amount of time. Granted, queueing would help but I still don't think it could be faster. Maybe that is something we could test on a smaller scale and consider transitioning to in the future if it is worth it. 

What I could try is Aaron's suggestion of using non-blocking channels in the internal bus and just drop unprocessed messages. This could potentially be even faster and, since we really only care about the metric when Prometheus scrapes once every second, there is no real reason to process every single incoming message anyway except those that get there in time to be scrapped. This is my opinion, however, and maybe this is not the functionality we want. I think, in this situation, we would want to be careful that all metrics are dropped evenly to avoid staleness problems of one metric. Maybe I'll try testing that approach today

pleimer commented:
I'll have to fix the coveralls badge for forked branches but it will work when this is merged
